### ISSUE TITLE
Undefined names have the potential to raise NameError at runtime

### ISSUE THREAD

**Issue by cclauss** [UIUC-PPL/charm4py#16]
[flake8](http://flake8.pycqa.org) testing of https://github.com/UIUC-PPL/charmpy on Python 3.7.0

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./examples/stencil3d/stencil3d_numba.py:174:9: F821 undefined name 'compute_kernel_fast'
        compute_kernel_fast(10, T, T)
        ^
./examples/stencil3d/stencil3d_numba.py:175:9: F821 undefined name 'constrainBC_fast'
        constrainBC_fast(T)
        ^
./examples/stencil3d/stencil3d_numba.py:176:9: F821 undefined name 'fillGhostData'
        fillGhostData(T, T, T, T, T, T, T)
        ^
./examples/stencil3d/stencil3d_numba.py:177:9: F821 undefined name 'processGhosts_fast'
        processGhosts_fast(T, LEFT, 1, 1, T)
        ^
./examples/stencil3d/stencil3d_numba.py:192:9: F821 undefined name 'constrainBC_fast'
        constrainBC_fast(self.temperature)
        ^
./examples/stencil3d/stencil3d_numba.py:225:9: F821 undefined name 'fillGhostData'
        fillGhostData(self.temperature, leftGhost, rightGhost, topGhost, bottomGhost, frontGhost, backGhost)
        ^
./examples/stencil3d/stencil3d_numba.py:236:9: F821 undefined name 'processGhosts_fast'
        processGhosts_fast(self.temperature, direction, width, height, gh)
        ^
./examples/stencil3d/stencil3d_numba.py:250:9: F821 undefined name 'constrainBC_fast'
        constrainBC_fast(self.temperature)
        ^
./examples/stencil3d/stencil3d_numba.py:281:9: F821 undefined name 'compute_kernel_fast'
        compute_kernel_fast(int(work), self.new_temperature, self.temperature)
        ^
./charmpy/chare.py:41:44: F821 undefined name 'Options'
        self._local = [i for i in range(1, Options.LOCAL_MSG_BUF_SIZE+1)]
                                           ^
./charmpy/chare.py:53:19: F821 undefined name 'CharmPyError'
            raise CharmPyError("Local msg buffer full. Increase LOCAL_MSG_BUF_SIZE")
                  ^
./charmpy/chare.py:103:28: F821 undefined name 'charm'
        if cond_str not in charm.wait_conditions:
                           ^
./charmpy/chare.py:105:13: F821 undefined name 'charm'
            charm.wait_conditions[cond_str] = cond_template
            ^
./charmpy/chare.py:107:29: F821 undefined name 'charm'
            cond_template = charm.wait_conditions[cond_str]
                            ^
./charmpy/chare.py:110:13: F821 undefined name 'charm'
            charm.threadMgr.pauseThread()
            ^
./charmpy/chare.py:113:9: F821 undefined name 'charm'
        charm.contribute(data, reducer_type, target, self)
        ^
./charmpy/chare.py:116:9: F821 undefined name 'charm'
        charm.contribute(data, Reducer.gather, target, self)
        ^
./charmpy/chare.py:116:32: F821 undefined name 'Reducer'
        charm.contribute(data, Reducer.gather, target, self)
                               ^
./charmpy/chare.py:120:9: F821 undefined name 'charm'
        charm.CkArraySend(self.thisProxy.aid, self.thisIndex, self.thisProxy.AtSync.ep, (b'', []))
        ^
./charmpy/chare.py:125:9: F821 undefined name 'charm'
        charm.lib.CkMigrate(self.thisProxy.aid, self.thisIndex, toPe)
        ^
./charmpy/chare.py:129:9: F821 undefined name 'charm'
        charm.threadMgr.depositFuture(fid, result)
        ^
./charmpy/chare.py:149:46: F821 undefined name 'charm'
            header[b'block'] = blockFuture = charm.createFuture()
                                             ^
./charmpy/chare.py:151:12: F821 undefined name 'Options'
        if Options.LOCAL_MSG_OPTIM and (cid in charm.chares) and (len(args) > 0):
           ^
./charmpy/chare.py:151:48: F821 undefined name 'charm'
        if Options.LOCAL_MSG_OPTIM and (cid in charm.chares) and (len(args) > 0):
                                               ^
./charmpy/chare.py:152:23: F821 undefined name 'charm'
            destObj = charm.chares[cid]
                      ^
./charmpy/chare.py:153:15: F821 undefined name 'charm'
        msg = charm.packMsg(destObj, args, header)
              ^
./charmpy/chare.py:154:9: F821 undefined name 'charm'
        charm.CkChareSend(cid, ep, msg)
        ^
./charmpy/chare.py:160:5: F821 undefined name 'charm'
    charm.CkContributeToChare(contributeInfo, proxy.cid)
    ^
./charmpy/chare.py:169:25: F821 undefined name 'charm'
        obj.thisProxy = charm.proxyClasses[MAINCHARE][obj.__class__](cid)
                        ^
./charmpy/chare.py:179:18: F821 undefined name 'charm'
        for m in charm.classEntryMethods[MAINCHARE][cls]:
                 ^
./charmpy/chare.py:181:23: F821 undefined name 'CharmPyError'
                raise CharmPyError("Unregistered entry method")
                      ^
./charmpy/chare.py:182:16: F821 undefined name 'Options'
            if Options.PROFILING:
               ^
./charmpy/chare.py:183:29: F821 undefined name 'profile_send_function'
                M[m.name] = profile_send_function(mainchare_proxy_method_gen(m.epIdx))
                            ^
./charmpy/chare.py:221:46: F821 undefined name 'charm'
            header[b'block'] = blockFuture = charm.createFuture()
                                             ^
./charmpy/chare.py:225:12: F821 undefined name 'Options'
        if Options.LOCAL_MSG_OPTIM and (elemIdx == charm._myPe) and (len(args) > 0):
           ^
./charmpy/chare.py:225:52: F821 undefined name 'charm'
        if Options.LOCAL_MSG_OPTIM and (elemIdx == charm._myPe) and (len(args) > 0):
                                                   ^
./charmpy/chare.py:226:23: F821 undefined name 'charm'
            destObj = charm.groups[proxy.gid]
                      ^
./charmpy/chare.py:227:15: F821 undefined name 'charm'
        msg = charm.packMsg(destObj, args, header)
              ^
./charmpy/chare.py:228:9: F821 undefined name 'charm'
        charm.CkGroupSend(proxy.gid, elemIdx, ep, msg)
        ^
./charmpy/chare.py:238:27: F821 undefined name 'charm'
        if get_ident() != charm.threadMgr.main_thread_id and ArrayMap not in C.mro():
                          ^
./charmpy/chare.py:239:31: F821 undefined name 'charm'
            creation_future = charm.createFuture()
                              ^
./charmpy/chare.py:241:15: F821 undefined name 'charm'
        msg = charm.packMsg(None, args, header)
              ^
./charmpy/chare.py:242:15: F821 undefined name 'charm'
        gid = charm.lib.CkCreateGroup(C.idx[GROUP], epIdx, msg)
              ^
./charmpy/chare.py:243:17: F821 undefined name 'charm'
        proxy = charm.groups[gid].thisProxy
                ^
./charmpy/chare.py:250:5: F821 undefined name 'charm'
    charm.CkContributeToGroup(contributeInfo, proxy.gid, proxy.elemIdx)
    ^
./charmpy/chare.py:259:19: F821 undefined name 'CharmPyError'
            raise CharmPyError("Only subclasses of Chare can be member of Group")
                  ^
./charmpy/chare.py:260:16: F821 undefined name 'charm'
        return charm.proxyClasses[GROUP][C].ckNew(args)
               ^
./charmpy/chare.py:264:25: F821 undefined name 'charm'
        obj.thisIndex = charm.myPe()
                        ^
./charmpy/chare.py:265:25: F821 undefined name 'charm'
        obj.thisProxy = charm.proxyClasses[GROUP][obj.__class__](gid)
                        ^
./charmpy/chare.py:266:31: F821 undefined name 'charm'
        obj._contributeInfo = charm.lib.initContributeInfo(gid, obj.thisIndex, CONTRIBUTOR_TYPE_GROUP)
                              ^
./charmpy/chare.py:276:24: F821 undefined name 'charm'
        entryMethods = charm.classEntryMethods[GROUP][cls]
                       ^
./charmpy/chare.py:279:23: F821 undefined name 'CharmPyError'
                raise CharmPyError("Unregistered entry method")
                      ^
./charmpy/chare.py:280:16: F821 undefined name 'Options'
            if Options.PROFILING:
               ^
./charmpy/chare.py:281:29: F821 undefined name 'profile_send_function'
                M[m.name] = profile_send_function(group_proxy_method_gen(m.epIdx))
                            ^
./charmpy/chare.py:315:15: F821 undefined name 'CharmPyError'
        raise CharmPyError("Dimensions of index " + str(idx) + " don't match array dimensions")
              ^
./charmpy/chare.py:325:46: F821 undefined name 'charm'
            header[b'block'] = blockFuture = charm.createFuture()
                                             ^
./charmpy/chare.py:329:12: F821 undefined name 'Options'
        if Options.LOCAL_MSG_OPTIM and (len(args) > 0):
           ^
./charmpy/chare.py:330:21: F821 undefined name 'charm'
            array = charm.arrays[proxy.aid]
                    ^
./charmpy/chare.py:333:15: F821 undefined name 'charm'
        msg = charm.packMsg(destObj, args, header)
              ^
./charmpy/chare.py:334:9: F821 undefined name 'charm'
        charm.CkArraySend(proxy.aid, elemIdx, ep, msg)
        ^
./charmpy/chare.py:346:19: F821 undefined name 'CharmPyError'
            raise CharmPyError("Bounds and number of dimensions for array cannot be empty in ckNew")
                  ^
./charmpy/chare.py:348:19: F821 undefined name 'CharmPyError'
            raise CharmPyError("Number of bounds should match number of dimensions")
                  ^
./charmpy/chare.py:357:27: F821 undefined name 'charm'
        if get_ident() != charm.threadMgr.main_thread_id:
                          ^
./charmpy/chare.py:358:31: F821 undefined name 'charm'
            creation_future = charm.createFuture()
                              ^
./charmpy/chare.py:361:15: F821 undefined name 'charm'
        msg = charm.packMsg(None, args, header)
              ^
./charmpy/chare.py:362:15: F821 undefined name 'charm'
        aid = charm.lib.CkCreateArray(C.idx[ARRAY], dims, epIdx, msg, map_gid)
              ^
./charmpy/chare.py:373:15: F821 undefined name 'charm'
        msg = charm.packMsg(None, args, {})
              ^
./charmpy/chare.py:374:9: F821 undefined name 'charm'
        charm.lib.CkInsert(proxy.aid, index, epIdx, onPE, msg)
        ^
./charmpy/chare.py:378:5: F821 undefined name 'charm'
    charm.CkContributeToArray(contributeInfo, proxy.aid, proxy.elemIdx)
    ^
./charmpy/chare.py:381:5: F821 undefined name 'charm'
    charm.lib.CkDoneInserting(proxy.aid)
    ^
./charmpy/chare.py:390:19: F821 undefined name 'CharmPyError'
            raise CharmPyError("Only subclasses of Chare can be member of Array")
                  ^
./charmpy/chare.py:391:16: F821 undefined name 'charm'
        return charm.proxyClasses[ARRAY][C].ckNew(dims, ndims, args, map)
               ^
./charmpy/chare.py:396:25: F821 undefined name 'charm'
        obj.thisProxy = charm.proxyClasses[ARRAY][obj.__class__](aid, len(obj.thisIndex))
                        ^
./charmpy/chare.py:399:31: F821 undefined name 'charm'
        obj._contributeInfo = charm.lib.initContributeInfo(aid, obj.thisIndex, CONTRIBUTOR_TYPE_ARRAY)
                              ^
./charmpy/chare.py:410:24: F821 undefined name 'charm'
        entryMethods = charm.classEntryMethods[ARRAY][cls]
                       ^
./charmpy/chare.py:413:23: F821 undefined name 'CharmPyError'
                raise CharmPyError("Unregistered entry method")
                      ^
./charmpy/chare.py:414:16: F821 undefined name 'Options'
            if Options.PROFILING:
               ^
./charmpy/chare.py:415:29: F821 undefined name 'profile_send_function'
                M[m.name] = profile_send_function(array_proxy_method_gen(m.epIdx))
                            ^
./charmpy/wait.py:143:17: F821 undefined name 'charm'
                charm.threadMgr.resumeThread(tid, None)
                ^
./charmpy/entry_method.py:34:9: F821 undefined name 'charm'
        charm.threadMgr.startThread(obj, self, args, header)
        ^
./charmpy/entry_method.py:47:9: F821 undefined name 'charm'
        charm.runningEntryMethod = self
        ^
./charmpy/entry_method.py:53:9: F821 undefined name 'charm'
        charm.runningEntryMethod = None
        ^
./charmpy/entry_method.py:75:30: F821 undefined name 'charm'
        self.__dict__.update(charm.entryMethods[ep].__dict__)
                             ^
./charmpy/threads.py:29:26: F821 undefined name 'charm'
            self.value = charm.threadMgr.pauseThread()
                         ^
./charmpy/threads.py:40:31: F821 undefined name 'charm'
        proxy_class = getattr(charm, self.proxy_class_name)
                              ^
./charmpy/threads.py:54:17: F821 undefined name 'charm'
                charm.threadMgr.resumeThread(self.tid, self.value)
                ^
./charmpy/threads.py:76:26: F821 undefined name 'Options'
        self.PROFILING = Options.PROFILING
                         ^
./charmpy/threads.py:115:23: F821 undefined name 'CharmPyError'
                raise CharmPyError("Migration of chares with active threads is not yet supported")
                      ^
./charmpy/threads.py:138:15: F821 undefined name 'CharmPyError'
        raise CharmPyError("Entry method '" + charm.mainThreadEntryMethod.C.__name__ + "." +
              ^
./charmpy/threads.py:138:47: F821 undefined name 'charm'
        raise CharmPyError("Entry method '" + charm.mainThreadEntryMethod.C.__name__ + "." +
                                              ^
./charmpy/threads.py:139:28: F821 undefined name 'charm'
                           charm.mainThreadEntryMethod.name +
                           ^
./charmpy/threads.py:171:13: F821 undefined name 'charm'
            charm.mainThreadEntryMethod.stopMeasuringTime()
            ^
./charmpy/threads.py:176:13: F821 undefined name 'charm'
            charm.mainThreadEntryMethod.startMeasuringTime()
            ^
93    F821 undefined name 'Options'
93
```

**Comment by juanjgalvez** [comment]
These are mostly globals that reference globals in other modules. CharmPy initialization is a bit complicated with the C++ library involved, so things are mostly initialized once the C++ library returns control at the beginning.

There are multiple modules with globals referencing globals in the `charm.py` module, and these are undefined until charm defines them at runtime. I guess I could define these names as `None` initially, maybe that would quiet flake8 it wouldn't really change anything.

I could maybe pass those references in a different manner. I'll look into this but also welcome suggestions.

**Comment by juanjgalvez** [comment]
Closed by 64b9175b29f5ceba134c2bc4d823b980b253ecc1
